src/platforms/common/px4_work_queue/WorkItem.cpp

```diff
@@ -0,0 +1,114 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2019 PX4 Development Team. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include "WorkItem.hpp"
+
+#include "WorkQueue.hpp"
+#include "WorkQueueManager.hpp"
+
+#include <px4_log.h>
+#include <drivers/drv_hrt.h>
+
+namespace px4
+{
+
+WorkItem::WorkItem(const wq_config &config)
+{
+#if WQ_ITEM_PERF
+	_perf_cycle_time = perf_alloc(PC_ELAPSED, "wq_cycle_run_time");
+	_perf_latency = perf_alloc(PC_ELAPSED, "wq_run_latency");
+	_perf_interval = perf_alloc(PC_INTERVAL, "wq_run_interval");
+#endif /* WQ_ITEM_PERF */
+
+	if (!Init(config)) {
+		PX4_ERR("init fail");
+	}
+}
+
+WorkItem::~WorkItem()
+{
+#if WQ_ITEM_PERF
+	perf_free(_perf_cycle_time);
+	perf_free(_perf_latency);
+	perf_free(_perf_interval);
+#endif /* WQ_ITEM_PERF */
+}
+
+bool WorkItem::Init(const wq_config &config)
+{
+	px4::WorkQueue *wq = work_queue_create(config);
+
+	if (wq != nullptr) {
+		_wq = wq;
+
+		return true;
+	}
+
+	return false;
+}
+
+void WorkItem::ScheduleNow()
+{
+	if (_wq != nullptr && !_queued) {
```

Comment:
There's a race conditon here between checking `_queued` and setting it.
If you move that logic into `WorkQueue::add`, you can protect it by its lock and don't have to do anything else (drop the volatile).
It would also make the code less error prone, as `_queued` is set at the same place, where the item is actually added to the queue.

I'd also take `_queued = false;` out of `pre_run` and add an explicit method for it, so it's easier to follow.

Another thing you can do is to add a `WorkItem::added_to_queue`, that sets the queue time and sets _queued.

Reply:
I was considering updating the `Queue` itself to prevent anything from being pushed more than once.

Reply:
Queue updated (in https://github.com/PX4/Firmware/pull/11570) to prevent duplicate publishes.